### ISSUE TITLE
Egress gateway doesn't work anymore on master with wireguard enabled

### ISSUE THREAD

**Issue by MrFreezeex** [cilium/cilium#24450]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

I tried latest master in my cluster with wireguard/egress gateway enabled. I noticed that traffic is dropped whenever egress gateway match something, disabling Cilium wireguard encryption resolved the issue.

While I am not exactly certain that traffic matching an egressgateway rule was encrypted in 1.13, I was able to establish connection with that version/settings. So something probably changed in Cilium's Wireguard that broke Egress gateway somehow?

### Cilium Version

master (latest commit as of now; commit: cb6505cee9f783393be99aef5d25a488aa74410e)

### Kernel Version

Linux xxxx 6.1.0-3-amd64 #1 SMP PREEMPT_DYNAMIC Debian 6.1.8-1 (2023-01-29) x86_64 GNU/Linux

### Kubernetes Version

Server Version: version.Info{Major:"1", Minor:"26", GitVersion:"v1.26.1+k3s1", GitCommit:"f10af367c3c96863c081ada4018e94e085c9404d", GitTreeState:"clean", BuildDate:"2023-01-26T00:59:34Z", GoVersion:"go1.19.5", Compiler:"gc", Platform:"linux/amd64"}

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by youngnick** [comment]
Thanks for logging this @MrFreezeex. Could I ask you to collect a sysdump for us, using https://docs.cilium.io/en/stable/operations/troubleshooting/#automatic-log-state-collection?

I'll ping this one to the folks working on Egress Gateway, but they're going to need a sysdump to be able to take a proper look.

Thanks again!

**Comment by MrFreezeex** [comment]
Sure here it is 
[cilium-sysdump-20230320-000017.zip](https://github.com/cilium/cilium/files/11012519/cilium-sysdump-20230320-000017.zip)

If that helps, I was testing a egress gateway policy named `egress-test-2` and this pod:
```
❯ k get ciliumendpoints.cilium.io -n default
NAME       ENDPOINT ID   IDENTITY ID   INGRESS ENFORCEMENT   EGRESS ENFORCEMENT   VISIBILITY POLICY   ENDPOINT STATE   IPV4        IPV6
my-shell   402           2168          <status disabled>     <status disabled>    <status disabled>   ready            10.0.1.98
```

I was using a dest cidr of 0.0.0.0/0 in the policy matching another node where it can't be scheduled. With wireguard enabled I had no connectivity and without wireguard on everything worked fine.

**Comment by pchaigno** [comment]
Hi @MrFreezeex! Thanks for the report and the sysdump!
When you say "traffic is dropped", what destination IP addresses were you trying to connect to?

> I was using a dest cidr of 0.0.0.0/0 in the policy matching another node where it can't be scheduled.

Do you mean the egress gateway node was selected to be a node other than the node where the `my-shell` pod is running?

**Comment by MrFreezeex** [comment]
> Do you mean the egress gateway node was selected to be a node other than the node where the `my-shell` pod is running?

Yes exactly I have a taint on a specific node preventing from running most pods there, and which I mainly use as an egress gateway for a few apps (and in this case the `my-shell` pod which was running on the "non-egress gateway node"). I was checking running apt update or by doing curl https://ifconfig.me, so trying to reach a random external service to my cluster.

**Comment by pchaigno** [comment]
I've checked the source code for that packet path and I am not seeing an explanation for this. I think we'll need to trace the connection to find out what I'm missing.

Would you be able to either:
- trace the connection on both the client and gateway nodes with `cilium monitor`, or
- enable Hubble and capture a new sysdump after retrying the failed connection?

If Hubble is enabled, then the sysdump will contain Hubble flows captured on the nodes, which should tell us if Cilium is dropping the packets and where they disappear.

**Comment by MrFreezeex** [comment]
Ah also not sure if it's in the sysdump or not but my "agent" node is a master (and the gateway node is not) so this might have an implication as it implicitly disable encrypted host2host communication.

**Comment by MrFreezeex** [comment]
> Note Hubble isn't specific to network policies. It can show all the same packet traces that `cilium monitor` is showing here, except it can dump it in JSON for easier consummation by other tools.

Ah hmm ok I though I needed to add some network policy so that hubble is able to capture things (even a dummy one that allow everything)

> > Ah also not sure if it's in the sysdump or not but my "agent" node is a master (and the gateway node is not) so this might have an implication as it implicitly disable encrypted host2host communication.
> 
> Did you try to reproduce with the gateway node on a worker node?

I only have two nodes on this cluster so not I can't really check if it works with two workers.

> I think you'll need to disable monitor aggregation (`monitor-aggregation: none`) to show all packet traces. We may be missing some here between `TRACE_TO_OVERLAY` and the point where SNAT happens.
> 
> If that doesn't show anything different, then you'll probably have to resort to https://github.com/cilium/pwru.


I checked with pwru directly and I can't see SNAT happening anywhere, so that might be it (note that I don't own the network outside of my nodes so I can't effictively see if the traffic is snat-ed or not but I am gussing that pwru should be able to catch it):
```
root@AGENT ~ # ./pwru --filter-dst-ip=1.0.0.1 --filter-port=80 --filter-proto=tcp --output-tuple
2023/03/27 14:11:38 Per cpu buffer size: 4096 bytes
2023/03/27 14:11:38 Attaching kprobes (via kprobe)...
1351 / 1351 [-----------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 102 p/s
2023/03/27 14:11:51 Attached (ignored 0)
2023/03/27 14:11:51 Listening for events..
               SKB    CPU          PROCESS                     FUNC
0xffff8b7006c8cae8      4           [curl]             ip_local_out 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]           __ip_local_out 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]                ip_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]             nf_hook_slow 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]    apparmor_ip_postroute 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]         ip_finish_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]       __ip_finish_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]        ip_finish_output2 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]     neigh_resolve_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]       __neigh_event_send 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]               eth_header 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]                 skb_push 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]         __dev_queue_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]      netdev_core_pick_tx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]        validate_xmit_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]       netif_skb_features 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]  passthru_features_check 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]     skb_network_protocol 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]  skb_csum_hwoffload_help 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]       validate_xmit_xfrm 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]      dev_hard_start_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]   skb_clone_tx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]        __dev_forward_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]       __dev_forward_skb2 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]         skb_scrub_packet 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]           eth_type_trans 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]               __netif_rx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]        netif_rx_internal 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]       enqueue_to_backlog 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl]      __netif_receive_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4           [curl] __netif_receive_skb_one_core 10.0.1.46:52474->1.0.0.1:80(tcp)
2023/03/27 14:12:12 Perf event ring buffer full, dropped 16 samples
0xffff8b7006c8cae8      4        [<empty>]             ip_local_out 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]           __ip_local_out 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]                ip_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]             nf_hook_slow 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]    apparmor_ip_postroute 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]         ip_finish_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]       __ip_finish_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]        ip_finish_output2 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]     neigh_resolve_output 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]               eth_header 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]                 skb_push 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]         __dev_queue_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]      netdev_core_pick_tx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]        validate_xmit_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]       netif_skb_features 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]  passthru_features_check 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]     skb_network_protocol 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]  skb_csum_hwoffload_help 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]       validate_xmit_xfrm 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]      dev_hard_start_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]   skb_clone_tx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]        __dev_forward_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]       __dev_forward_skb2 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]         skb_scrub_packet 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]           eth_type_trans 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]               __netif_rx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]        netif_rx_internal 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4        [<empty>]       enqueue_to_backlog 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]      __netif_receive_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4] __netif_receive_skb_one_core 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]             tcf_classify 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]          skb_do_redirect 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]           __bpf_redirect 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]         __dev_queue_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]             tcf_classify 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]      netdev_core_pick_tx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]        validate_xmit_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]       netif_skb_features 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]     skb_network_protocol 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]  skb_csum_hwoffload_help 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]       validate_xmit_xfrm 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]      dev_hard_start_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4]    skb_tunnel_check_pmtu 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff8b7006c8cae8      4    [ksoftirqd/4] iptunnel_handle_offloads 10.0.1.46:52474->1.0.0.1:80(tcp)
```

```
root@GATEWAY:~# ./pwru --filter-dst-ip=1.0.0.1 --filter-port=80 --filter-proto=tcp --output-tuple
2023/03/27 07:09:49 Per cpu buffer size: 4096 bytes
2023/03/27 07:09:49 Attaching kprobes (via kprobe)...
1351 / 1351 [------------------------------------------------------------------------------------------------------------------------------------------------------------------] 100.00% 95 p/s
2023/03/27 07:10:03 Attached (ignored 0)
2023/03/27 07:10:03 Listening for events..
               SKB    CPU          PROCESS                     FUNC
0xffff9ee2de56ff00      0        [<empty>]         tcp4_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0        [<empty>] __skb_gro_checksum_complete 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0        [<empty>]          tcp_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0        [<empty>]   skb_defer_rx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]         tcp4_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]          tcp_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]   skb_defer_rx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]             tcf_classify 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]          skb_do_redirect 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]           __bpf_redirect 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]         __dev_queue_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      netdev_core_pick_tx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]        validate_xmit_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]       netif_skb_features 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]  passthru_features_check 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]     skb_network_protocol 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]       validate_xmit_xfrm 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      dev_hard_start_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]   skb_clone_tx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]        __dev_forward_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]       __dev_forward_skb2 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]         skb_scrub_packet 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]           eth_type_trans 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]               __netif_rx 10.0.1.46:52474->1.0.0.1:80(tcp)
2023/03/27 07:11:10 Perf event ring buffer full, dropped 51 samples
0xffff9ee2de56ff00      0        [<empty>]         tcp4_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0        [<empty>] __skb_gro_checksum_complete 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0        [<empty>]          tcp_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0        [<empty>]   skb_defer_rx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]         tcp4_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]          tcp_gro_receive 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]   skb_defer_rx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]             tcf_classify 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      skb_ensure_writable 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]          skb_do_redirect 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]           __bpf_redirect 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]         __dev_queue_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      netdev_core_pick_tx 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]        validate_xmit_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]       netif_skb_features 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]  passthru_features_check 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]     skb_network_protocol 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]       validate_xmit_xfrm 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]      dev_hard_start_xmit 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]   skb_clone_tx_timestamp 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]        __dev_forward_skb 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]       __dev_forward_skb2 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]         skb_scrub_packet 10.0.1.46:52474->1.0.0.1:80(tcp)
0xffff9ee2de56ff00      0    [ksoftirqd/0]           eth_type_trans 10.0.1.46:52474->1.0.0.1:80(tcp)
```

**Comment by pchaigno** [comment]
I'm not seeing anything in the pwru output that suggests the packet was dropped :thinking: Do you see it leave on the gateway's native device with tcpdump? Or maybe on the gateway's wireguard interface?

**Comment by MrFreezeex** [comment]
Here is the tcpdump:
```
root@AGENT ~ # tcpdump -i any host 1.0.0.1 -n
tcpdump: data link type LINUX_SLL2
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on any, link-type LINUX_SLL2 (Linux cooked v2), snapshot length 262144 bytes
14:41:28.058132 lxc49308e1cc171 In  IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114493048 ecr 0,nop,wscale 7], length 0
14:41:28.058149 cilium_vxlan Out IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114493048 ecr 0,nop,wscale 7], length 0
14:41:29.072306 lxc49308e1cc171 In  IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114494062 ecr 0,nop,wscale 7], length 0
14:41:29.072323 cilium_vxlan Out IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114494062 ecr 0,nop,wscale 7], length 0
```

```
root@GATEWAY:~# tcpdump -i any host 1.0.0.1 -n
tcpdump: data link type LINUX_SLL2
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on any, link-type LINUX_SLL2 (Linux cooked v2), snapshot length 262144 bytes
07:40:26.171092 cilium_vxlan P   IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114493048 ecr 0,nop,wscale 7], length 0
07:40:26.171103 cilium_net Out IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114493048 ecr 0,nop,wscale 7], length 0
07:40:26.171106 cilium_host In  IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114493048 ecr 0,nop,wscale 7], length 0
07:40:26.171127 ens3  Out IP 107.189.8.245.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114493048 ecr 0,nop,wscale 7], length 0
07:40:26.176272 ens3  In  IP 1.0.0.1.80 > 107.189.8.245.49080: Flags [S.], seq 1246766322, ack 2050669321, win 65160, options [mss 1400,sackOK,TS val 3014951774 ecr 2114493048,nop,wscale 13], length 0
07:40:26.176292 cilium_wg0 Out IP 1.0.0.1.80 > 10.0.1.46.49080: Flags [S.], seq 1246766322, ack 2050669321, win 65160, options [mss 1400,sackOK,TS val 3014951774 ecr 2114493048,nop,wscale 13], length 0
07:40:27.185277 cilium_vxlan P   IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114494062 ecr 0,nop,wscale 7], length 0
07:40:27.185288 cilium_net Out IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114494062 ecr 0,nop,wscale 7], length 0
07:40:27.185292 cilium_host In  IP 10.0.1.46.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114494062 ecr 0,nop,wscale 7], length 0
07:40:27.185312 ens3  Out IP 107.189.8.245.49080 > 1.0.0.1.80: Flags [S], seq 2050669320, win 64860, options [mss 1380,sackOK,TS val 2114494062 ecr 0,nop,wscale 7], length 0
07:40:27.190490 ens3  In  IP 1.0.0.1.80 > 107.189.8.245.49080: Flags [S.], seq 1246766322, ack 2050669321, win 65160, options [mss 1400,sackOK,TS val 3014952788 ecr 2114493048,nop,wscale 13], length 0
07:40:27.190537 cilium_wg0 Out IP 1.0.0.1.80 > 10.0.1.46.49080: Flags [S.], seq 1246766322, ack 2050669321, win 65160, options [mss 1400,sackOK,TS val 3014952788 ecr 2114493048,nop,wscale 13], length 0
07:40:28.239463 ens3  In  IP 1.0.0.1.80 > 107.189.8.245.49080: Flags [S.], seq 1246766322, ack 2050669321, win 65160, options [mss 1400,sackOK,TS val 3014953837 ecr 2114493048,nop,wscale 13], length 0
07:40:28.239503 cilium_wg0 Out IP 1.0.0.1.80 > 10.0.1.46.49080: Flags [S.], seq 1246766322, ack 2050669321, win 65160, options [mss 1400,sackOK,TS val 3014953837 ecr 2114493048,nop,wscale 13], length 0
```

So it goes like that: agent -> (via vxlan) gateway -> (snated correctly) 1.0.0.1 -> gateway -> (via wg0; never received) agent apparently according to the tcpdump

**Comment by pchaigno** [comment]
Ahah! We're getting somewhere (contrary to those SYN+ACKs).

So that packet path can explain the drops because the SYN travels via cilium_vxlan and the SYN+ACK via cilium_wg0 which means we have a path asymmetry. Path asymmetries can confuse all sort of things (most often Linux's conntrack) and cause packets to be dropped.

What was unexpected to me here is that the SYN+ACK is sent through cilium_wg0. Given the source IP address is not part of the cluster, I wouldn't have expected that to happen.

It happens because of a bug in this logic: https://github.com/cilium/cilium/blob/d8799e686d087c41c3985e5d53430dab1d831144/bpf/lib/wireguard.h#L90-L93
Basically, we check that the source IP address is not a pod (traffic from e.g. hostns to pod should only be encrypted if node encryption is enabled). We incorrectly do that using the ipcache map, but the ipcache map will have a fallback entry 0.0.0.0/0 matching all (and it can also match because of CIDR policies), thus we find an entry with `src->sec_label` different from HOST_ID and send the packet to be encrypted.

I'll send a fix.

**Comment by pchaigno** [comment]
@MrFreezeex Would you be able to try the fix at https://github.com/cilium/cilium/pull/24586? :pray:

**Comment by pchaigno** [comment]
Thanks @MrFreezeex for the report! We probably wouldn't have solved this so quickly without your prompt replies and many details! :pray:

**Comment by MrFreezeex** [comment]
> Thanks @MrFreezeex for the report! We probably wouldn't have solved this so quickly without your prompt replies and many details! pray

And thanks for the super quick fix and responsiveness :D! I actually tested stuff on master following some issues I had with egressgateway on the latest stable release, and most of them were already fixed in master (like cilium not updating his internal state from k8s one, this mostly IIUC https://github.com/cilium/cilium/pull/24034) so many thanks to the CIlium team for that as well :pray:.

I was also testing a patch of my own here https://github.com/cilium/cilium/pull/24449 (shameless plug I know :p, no pressure for reviewing though it's relatively recent and all).

**Comment by pchaigno** [comment]
> some issues I had with egressgateway on the latest stable release, and most of them were already fixed in master (like cilium not updating his internal state from k8s one, this mostly IIUC #24034) so many thanks to the CIlium team for that as well pray.

If you're hitting some of this in production, don't hesitate to drop a comment on the PRs for us to consider backporting to the Cilium release you are running. I can't promise we will backport (it's always a tradeoff between risk and gain), but it's definitely useful input.

> I was also testing a patch of my own here #24449 (shameless plug I know :p, no pressure for reviewing though it's relatively recent and all).

I've let the assigned reviewer know about this, but they are currently at IETF so it may wait until next week.
